### ISSUE TITLE
Decouple storage controller processing from the coordinator

### ISSUE THREAD

**Issue by teskje** [MaterializeInc/materialize#29486]
The storage controller is currently mostly implemented as a library whose usage follows a "ready/process" pattern. The coordinator's message loop `select`s on the `ready` method, along with other things it waits on. When `ready` resolves, that signals that the storage controller has work to do, so the coordinator invokes its `process` method. `process` then performs any work that has queued up in the compute controller.

The key issue is that a significant portion of the storage controller's work is executed on the coordinator's main thread that also processes user requests. Some of this work (e.g., `allow_writes`, `create_collections`, various `alter_...` methods) make synchronous calls to persist and are therefore blocking on network traffic. And even work that does not block still uses CPU time and can therefore interfere with the processing of user requests, reducing latency.

Similarly to the compute controller (#27656) we should move the core storage controller logic to a separate task that's connected to the coordinator through channels.

The main challenge here is figuring out (a) what assumptions the coordinator currently makes about the state of the system when a storage controller call returns and (b) how we can relax those to reduce the amount of synchronization between the two. For example:
* Will we run into consistency issues when the coordinator does not wait for storage collections to be created after invoking `create_collections`?
* When inserting into a table, we presumably need to wait for the storage controller to acknowledge the append before we can report the insert as successful to the user.
